### ISSUE TITLE
Ability to upload stubs to stubo (instead of stubo downloading from servers)

### ISSUE THREAD

**Issue by sponte** [Stub-O-Matic/stubo-app#96]
Can you please add ability to upload stubs to stubo directly instead of telling stubo where stubs live?

We are in a process of moving stub files into our main repository and would like to version them along with code. Once we have ability to upload compressed stubs (jars or zips) we'll be able to manage stubs in a much more efficient fashion. We will also avoid lots of networking configuration required for stubo to pull required content.

**Comment by rusenask** [comment]
Yes, it's on my todo list, it will happen to modules as well. We will probably stick with zip format and try to deprecate jars eventually.

**Comment by rusenask** [comment]
@sponte - looked into this a bit more, however it poses some challenges since import function just calls stubo itself and it isn't the best approach. What would you think about a separate client application that could be preinstalled on Jenkins or on any other machine in which you could just specify .zip file location and it would then import that file (with scenario and stubs) into stubo?

**Comment by sponte** [comment]
@rusenask I think this would introduce unnecessary complexity by breaking stubo into multiple services based application. Could we add a new endpoint within stubo that would accept zip, expand and create temporary web server for stubo to import from?

**Comment by johnecobo** [comment]
Are you suggesting that one is able to POST to stubo with the zipped command and stub files as the HTTP body?

**Comment by sponte** [comment]
Kind of, I'm proposing Stubo to allow upload of such zip files as
MultiPart - standard way of uploading files to HTTP servers. This way
stubo would not have to be concerned with firewall issues, it could
even live outside of the network if it was properly secured.

On 01/09/2015, John Cobo <notifications@github.com> wrote:
> Are you suggesting that one is able to POST to stubo with the zipped command
> and stub files as the HTTP body?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Stub-O-Matic/stubo-app/issues/96#issuecomment-136750085


-- 
Kind regards,
Stanislaw Wozniak

**Comment by johnecobo** [comment]
If I understand you correctly, such an upload would only be useful for
manual loading. I'd encourage people to use Stubo in a fully automated way
- think overnight batch test runs.

**Comment by sponte** [comment]
This does not fit our use case. We would like to upload versions stubs from
within source code on the fly. Think same stub but different versions when
run from different branches or tags.

On Tuesday, 1 September 2015, John Cobo <notifications@github.com> wrote:

> If I understand you correctly, such an upload would only be useful for
> manual loading. I'd encourage people to use Stubo in a fully automated way
> - think overnight batch test runs.
>
> On 1 September 2015 at 18:00, Stanislaw Wozniak <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> wrote:
>
> > Kind of, I'm proposing Stubo to allow upload of such zip files as
> > MultiPart - standard way of uploading files to HTTP servers. This way
> > stubo would not have to be concerned with firewall issues, it could
> > even live outside of the network if it was properly secured.
> >
> > On 01/09/2015, John Cobo <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> > > Are you suggesting that one is able to POST to stubo with the zipped
> > command
> > > and stub files as the HTTP body?
> > >
> > > ---
> > > Reply to this email directly or view it on GitHub:
> > >
> >
> https://github.com/Stub-O-Matic/stubo-app/issues/96#issuecomment-136750085
> >
> >
> > --
> > Kind regards,
> > Stanislaw Wozniak
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/Stub-O-Matic/stubo-app/issues/96#issuecomment-136795691
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Stub-O-Matic/stubo-app/issues/96#issuecomment-136804052>
> .
>


-- 
Kind regards,
Stanislaw Wozniak

**Comment by rusenask** [comment]
They could put archived scenario content in the body of the HTTP request, which will have content-type of application/octet-stream. However, not sure how they would initiate such request.

**Comment by sponte** [comment]
Actually we would do a Multipart POST similar to http://www.codejava.net/java-se/networking/upload-files-by-sending-multipart-request-programmatically. This is a standard way of uploading files - equivalent to posting through a web browser. 

Kind regards,
Stanislaw Wozniak

* Sent from my iPhone

**Comment by johnecobo** [comment]
Excellent, the multi-part POST makes sense to me.

**Comment by rusenask** [comment]
Another solution would be to host that zip file on your machine (the same one as you would be uploading it from) and let stubo download it. Have you thought about it?

**Comment by sponte** [comment]
Again this would require stubo to have network access to the host which would require network / firewall changes in some cases.

**Comment by rusenask** [comment]
How you would be uploading file if you don't have network access?

**Comment by sponte** [comment]
Firewall would only need to be opened one way, from client network to stubo host rather than opening stubo host's netowork to arbitrary number of hosts within client's network.

**Comment by JohnFDavenport** [comment]
Closing this is as it has been superseded by SpectoLabs/mirage#10
